library/init/category/reader.lean

```diff
@@ -0,0 +1,80 @@
+/-
+Copyright (c) 2017 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sebastian Ullrich
+
+The reader monad transformer for passing immutable state.
+-/
+
+prelude
+import init.category.lift init.category.id
+universes u v w
+
+structure reader_t (r : Type u) (m : Type u → Type v) (α : Type u) : Type (max u v) :=
+(run : r → m α)
+
+@[reducible] def reader (r : Type u) := reader_t r id
+
+namespace reader_t
+section
+  variable  {r : Type u}
+  variable  {m : Type u → Type v}
+  variable  [monad m]
+  variables {α β : Type u}
+
+  protected def read : reader_t r m r :=
+  ⟨pure⟩
+
+  protected def pure (a : α) : reader_t r m α :=
+  ⟨λ r, pure a⟩
+
+  protected def bind (x : reader_t r m α) (f : α → reader_t r m β) : reader_t r m β :=
+  ⟨λ r, do a ← x.run r,
+           (f a).run r⟩
+
+  instance : monad (reader_t r m) :=
+  { pure := @reader_t.pure _ _ _, bind := @reader_t.bind _ _ _ }
+
+  protected def lift (a : m α) : reader_t r m α :=
+  ⟨λ r, a⟩
+
+  instance (m) [monad m] : has_monad_lift m (reader_t r m) :=
+  ⟨@reader_t.lift r m _⟩
+
+  protected def monad_map {r m m'} [monad m] [monad m'] {α} (f : Π {α}, m α → m' α) : reader_t r m α → reader_t r m' α :=
+  λ x, ⟨λ r, f (x.run r)⟩
+
+  instance (r m m') [monad m] [monad m'] : monad_functor m m' (reader_t r m) (reader_t r m') :=
+  ⟨@reader_t.monad_map r m m' _ _⟩
+end
+end reader_t
+
+
+/-- A specialization of `monad_lift` to `reader_t` that allows `r` to be inferred. -/
+class monad_reader_lift (r : out_param (Type u)) (m : out_param (Type u → Type v)) (n : Type u → Type w) :=
```

Comment:
Is this a trick/workaround to allow us to use `read` even when the environment `r` is not known? That is, if you have used `[has_monad_lift m (reader_t r m)]` instead of `[monad_reader_lift r m n]` to define `read`, then `r` would have to be known to trigger type class resolution. If this is the only motivation for the existence of `monad_reader_lift`, we should document it (describe the problem being solved, use examples, etc).